### PR TITLE
#165305756 Fix Password Reset Feature

### PR DESCRIPTION
**Pull request by Ogutu-Brian**
#### Description
Currently, when a user forgets his/her password, a password reset link is sent to their email accounts. However, on click of the link, they would be directed to a page that does not exist and when they make mistakes as they change their passwords, the error messages they receive are not descriptive. This PR makes the password reset link useful to the user and provides them with a form to change their passwords and also given more descriptive error messages.

#### Type of change
- [x] Bug fix (nonbreaking change which fixes an issue)
- [ ] New feature (nonbreaking change which fixes an issue)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

#### How Has This Been Tested?
Please describe the tests that you ran to verify your changes
- [ ] End to end
- [x] Integration

#### Checklist:
- [x] My code follows the style guide for this project
- [x] I have linted my code prior to submission
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

#### Pivotal Tracker
[165305756](https://www.pivotaltracker.com/story/show/165305756)

### CODE REVIEW

File: authors/apps/authentication/serializers.py [hunk 1]
```diff
@@ -214,55 +215,129 @@ class SocialAuthSerializer(serializers.Serializer):
...
+        policy = PasswordPolicy.from_names(
+            uppercase=2,
+            numbers=2,
+            special=2
+        )
+        if policy.test(data.get("password")):
+            stats = PasswordStats(data.get("password"))
+            if stats.letters_uppercase < 2:
+                raise serializers.ValidationError({
+                    "password": "Your password should have a minimum of 2 uppercase letters"
```
Comment (IssaIan) [inline authors/apps/authentication/serializers.py]: Here the minimum uppercase letters are one as @Philipotieno had set in the validations. Please confirm that with him.

File: authors/apps/authentication/serializers.py [hunk 2]
```diff
@@ -214,55 +215,129 @@ class SocialAuthSerializer(serializers.Serializer):
...
+                })
+            elif stats.numbers < 2:
+                raise serializers.ValidationError({
+                    "password": "Your password should have a minimum of 2  numbers"
+                })
+            elif stats.special_characters < 2:
```
Comment (IssaIan) [inline authors/apps/authentication/serializers.py]: Same case here. Work closely with @Philipotieno to be uniform in this validations.

File: authors/apps/authentication/serializers.py [hunk 3]
```diff
@@ -214,55 +215,129 @@ class SocialAuthSerializer(serializers.Serializer):
...
+
+    def get_password_field_erros(self, data):
+        """
+        Captures null, blank and empty fields for password
+        and password confirm
+        """
+        if not data.get("password") and not data.get("password_confirm"):
+            raise serializers.ValidationError({
+                "password": "Please provide your password",
+                "password_confirm": "Please provide your confirmation password"
```
Comment (IssaIan) [inline authors/apps/authentication/serializers.py]: I think `Please provide your confirmation password` can be made a better message. For instance `Please confirm your password`

File: authors/apps/authentication/serializers.py [hunk 4]
```diff
@@ -214,55 +215,129 @@ class SocialAuthSerializer(serializers.Serializer):
...
+                    "password": "Your password should have a minimum of 2 special characters"
+                })
+        return data
+
+    def get_password_field_erros(self, data):
+        """
+        Captures null, blank and empty fields for password
+        and password confirm
+        """
+        if not data.get("password") and not data.get("password_confirm"):
```
Comment (IssaIan) [inline authors/apps/authentication/serializers.py:331]: Isn't this validation repeated here:
```
elif not data.get("password"):
            raise serializers.ValidationError({
                "password": "passwords did not match"
                "password": "Please provide your password"
            })
        elif not data.get("password_confirm"):
            raise serializers.ValidationError({
                "password_confirm": "Please confirm your password"
```
Why have this validation here? The latter validations still check for this. If one field is empty, an error message will still be returned. Isn't capturing for both of them and then afterwards capturing for each one of the a little bit redundant??? Please enlighten me here.

**Review by IssaIan — CHANGES_REQUESTED** [review]

File: authors/apps/authentication/serializers.py (hunk 4 above)
Comment (IssaIan) [reply authors/apps/authentication/serializers.py:331]: Oh I get it now @Ogutu-Brian why there are those validations.

Comment (IssaIan) [reply authors/apps/authentication/serializers.py:331]: But then remember to add tests for the first edge case there. As they are different cases which need to be tested.

File: authors/apps/authentication/serializers.py (hunk 1 above)
Comment (Ogutu-Brian) [reply authors/apps/authentication/serializers.py]: Okay, let me do that

File: authors/apps/authentication/serializers.py (hunk 2 above)
Comment (Ogutu-Brian) [reply authors/apps/authentication/serializers.py]: Let me reach out to him

File: authors/apps/authentication/serializers.py (hunk 3 above)
Comment (Ogutu-Brian) [reply authors/apps/authentication/serializers.py]: Let me update that

File: authors/apps/authentication/serializers.py (hunk 4 above)
Comment (Ogutu-Brian) [reply authors/apps/authentication/serializers.py]: Hello @IssaIan ,
The two cases are already captured in the tests. Check in the test_password_reset.py file

File: authors/apps/authentication/serializers.py (hunk 3 above)
Comment (Ogutu-Brian) [reply authors/apps/authentication/serializers.py]: Updated

File: authors/apps/authentication/serializers.py (hunk 2 above)
Comment (Ogutu-Brian) [reply authors/apps/authentication/serializers.py]: This is the policy we have settle for:
```
3 lowercase
2 uppercase
1 special character
2 digits
```
I have made the necessary changes

File: authors/apps/authentication/tests/test_password_reset.py [hunk 5]
```diff
@@ -71,10 +71,25 @@ def test_null_email(self):
...
+        reset link
+        """
+        response = self.password_reset()
+        self.assertEqual(
+            response.status_code,
+            status.HTTP_200_OK
+        )
+        self.assertEqual(
+            response.data.get("data")[0].get("message"),
+            "An email has been sent to your email address to reset your password"
```
Comment (IssaIan) [inline authors/apps/authentication/tests/test_password_reset.py]: This message could also be better. For instance:
```
A password reset message was sent to your email address. Please click the link in that message to reset your password. 
```

File: authors/apps/authentication/serializers.py (hunk 1 above)
Comment (Ogutu-Brian) [reply authors/apps/authentication/serializers.py]: Done. However, we have set for a minimum of 2 uppercase

File: authors/apps/authentication/serializers.py (hunk 4 above)
Comment (Ogutu-Brian) [reply authors/apps/authentication/serializers.py]: I've added the tests

**Review by IssaIan — CHANGES_REQUESTED** [review]

File: authors/apps/authentication/serializers.py [hunk 6]
```diff
@@ -214,55 +215,129 @@ class SocialAuthSerializer(serializers.Serializer):
...
+    def validate(self, data):
+        email = data.get("email")
+        if not email:
+            raise serializers.ValidationError({
+                "email": "Please provide your email address"
+            })
         try:
-            EmailValidator(value)
+            EmailValidator(email)
         except ValidationError as e:
```
Comment (IssaIan) [inline authors/apps/authentication/serializers.py:237]: This edge case here is not tested in your code...Try testing for an invalid email to improve the tests.

**Review by IssaIan — CHANGES_REQUESTED** [review]

File: authors/apps/authentication/tests/test_password_reset.py (hunk 5 above)
Comment (Ogutu-Brian) [reply authors/apps/authentication/tests/test_password_reset.py]: I've update it

File: authors/apps/authentication/serializers.py (hunk 6 above)
Comment (Ogutu-Brian) [reply authors/apps/authentication/serializers.py:237]: Hello @IssaIan ,
It is already tested. It is the first test in` test_password_reset.py` file

**Review by IssaIan — APPROVED** [review]

**Review by a-braham — APPROVED** [review]
LGTM

**Review by Amoswachira — APPROVED** [review]
